lib/core/test/unit/Cardano/Pool/DB/Properties.hs

```diff
@@ -0,0 +1,146 @@
+import qualified Data.Map.Strict as Map
+
+-- | Provide a DBLayer to a Spec that requires it. The database is initialised
+-- once, and cleared with 'cleanDB' before each test.
+withDB :: IO (DBLayer IO) -> SpecWith (DBLayer IO) -> Spec
+withDB create = beforeAll create . beforeWith (\db -> cleanDB db $> db)
+
+-- | Set up a DBLayer for testing, with the command context, and the logging
+-- variable.
+newMemoryDBLayer :: IO Configuration -> IO (DBLayer IO)
+newMemoryDBLayer conf = snd . snd <$> (newMemoryDBLayer' conf)
+
+newMemoryDBLayer'
+    :: IO Configuration -> IO (TVar [LogObject Text], (SqliteContext, DBLayer IO))
+newMemoryDBLayer' testingLogConfig = do
+    logConfig <- testingLogConfig
+    logVar <- newTVarIO []
+    (logVar, ) <$> newDBLayer logConfig (traceInTVarIO logVar) Nothing
+
+properties :: SpecWith (DBLayer IO)
+properties = do
+    describe "Stake Pool properties" $ do
+        it "putPoolProduction . readPoolProduction yields expected results"
+            (property . prop_putReadPoolProduction)
+        it "putPoolProduction with already put slot yields error"
+            (property . prop_putSlotTwicePoolProduction)
+
+{-------------------------------------------------------------------------------
+                                    Properties
+-------------------------------------------------------------------------------}
+
+-- | Can read put pool production
+prop_putReadPoolProduction
+    :: DBLayer IO
+    -> StakePoolsFixture
+    -> Property
+prop_putReadPoolProduction db (StakePoolsFixture pairs) =
+    monadicIO (setup >>= prop)
+  where
+    epochs = nubOrd $ map (epochNumber . snd) pairs
+    setup = liftIO $ do
+        cleanDB db
+        db' <- MVar.newDBLayer
+        cleanDB db'
+        pure db'
+    prop db' = liftIO $ do
+        forM_ pairs $ \(pool, slot) ->
+            unsafeRunExceptT $ putPoolProduction db slot pool
+        forM_ pairs $ \(pool, slot) ->
+            unsafeRunExceptT $ putPoolProduction db' slot pool
+        forM_ epochs $ \(epoch) -> do
+            res' <- readPoolProduction db' epoch
+            readPoolProduction db epoch `shouldReturn` res'
+
+-- | Cannot put pool production with already put slot
+prop_putSlotTwicePoolProduction
+    :: DBLayer IO
+    -> StakePoolsFixture
+    -> Property
+prop_putSlotTwicePoolProduction db (StakePoolsFixture pairs) =
+    monadicIO (setup >>= prop)
+  where
+    setup = liftIO $ do
+        cleanDB db
+        db' <- MVar.newDBLayer
+        cleanDB db'
+        pure db'
+    prop db' = liftIO $ do
+        forM_ pairs $ \(pool, slot) -> do
+            let err = ErrSlotAlreadyExists slot
+            runExceptT (putPoolProduction db' slot pool) `shouldReturn` Right ()
+            runExceptT (putPoolProduction db' slot pool) `shouldReturn` Left err
+            runExceptT (putPoolProduction db slot pool) `shouldReturn` Right ()
+            --runExceptT (putPoolProduction db slot pool) `shouldReturn` Left err
```

Comment:
unfortunately, uncommenting the above gives rise to test failure. Which is strange, as we have : 
```
PoolProduction
    poolProductionEpochNumber    Word64       sql=epoch_number
    poolProductionSlotNumber     Word16       sql=slot_number
    poolProductionPoolId         W.PoolId     sql=pool_id
    Primary
        poolProductionEpochNumber
        poolProductionSlotNumber
    deriving Show Generic
```
and we use in Sqlite implementation : 
```
        { putPoolProduction = \s@(W.SlotId epoch slot) pool ->
            ExceptT $ runQuery $ handleConstraint (ErrSlotAlreadyExists s) $
            insert_ (PoolProduction epoch slot pool)
```
:thinking:

Reply:
Query logging is really helpful to debug this (it was a persistent DSL syntax issue).
The test database setup functions should be designed so that it's really easy to enable query logging for a test case - currently this requires modifying the code.